docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,821 @@
+
+### 4.2 Existing data: `UserManifest`
+
+Remove `last_processed_message` field.
+Modify `workspaces` field, i.e. the `WorkspaceEntry` nested type:
+
+```json5
+{
+    "label": "WorkspaceEntry",
+    "fields": [
+        {
+            "name": "id",
+            "type": "VlobID"
+        },
+        {
+            "name": "name",
+            "type": "EntryName"
+        }
+        // Removed fields:
+        // - key
+        // - encryption_revision
+        // - encrypted_on
+        // - role_cached_on
+        // - role_cache_value
+    ]
+}
+```
+
+> **Note:**
+>
+> `role_cached_on` and `role_cache_value` are not directly related to this RFC,
+> but given they are no longer in use in Parsec v3, we also remove them.
+
+> **TODO:**
+>
+> Should we rename workspace name into label ?
+>
+> We have "Human handle label" and "Device label", the "label" refers to the
+> fact they are purely display information (i.e. not used for unique identification).
+>
+> So given the idea is similar here (this field can even be modified by the user !),
+> it makes sens to rename it to "workspace label".
+>
+> On the other hand keeping "workspace name" is less work, and the fact this name
+> can be modified where the human handle / device label cannot may justify this
+> is a different concept...
+
+### 4.3 New certificate: `realm_key_rotation_certificate`
+
+```json5
+{
+    "label": "RealmKeyRotationCertificate",
+    "type": "realm_key_rotation_certificate",
+    "other_fields": [
+        {
+            "name": "author",
+            "type": "DeviceID"
+        },
+        {
+            "name": "timestamp",
+            "type": "DateTime"
+        },
+        {
+            "name": "realm_id",
+            "type": "VlobID"
+        },
+        {
+            "name": "key_index",
+            "type": "Index"
+        },
+        {
+            "name": "encryption_algorithm",
+            "type": "SecretKeyAlgorithm"
+        },
+        {
+            "name": "hash_algorithm",
+            "type": "HashAlgorithm"
+        },
+        {
+            // Empty message encrypted with the key, this is used to ensure the
+            // the key retrieved in the keys-bundle hasn't been tempered (given
+            // each key rotation author generates a new keys-bundle from the
+            // previous bundle and his one additional key).
+            //
+            // Note since the message is empty, we are doing here AEAD-without-data
+            // given we only care of the MAC part.
+            "name": "key_canary",
+            "type": "Bytes"
+        }
+    ]
+}
+```
+
+When processing this certificate (in addition to regular access checks), the
+client must make sure the `key_index` field strictly follow the value of the
+previous key rotation certificate for this realm. (i.e. we reject the
+certificate if there is a gap in key indexes)
+
+`SecretKeyAlgorithm` has one possible values: `XSALSA20-POLY1305`.
```

Comment:
Since this is new stuff we could rely on algorithm that has been audited in `RustCrypto` ?

That way we could plan to remove `sodiumoxide` in the dependency of `libparsec_crypto` that would simplify its implementation, dependencies specification, tests and packaging in the long run

Reply:
> Since this is new stuff we could rely on algorithm that has been audited in RustCrypto ?

Yes that's the idea: switch to ietf's chacha20-poly1305 which is much more scrutinized as it is used for TLS (for instance RustCrypto implementation has already been audited)